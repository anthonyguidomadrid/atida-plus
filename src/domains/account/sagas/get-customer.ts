import type { PayloadAction } from '@reduxjs/toolkit'
import { AxiosResponse } from 'axios'
import type { SagaIterator } from 'redux-saga'
import {
  all,
  call,
  getContext,
  put,
  takeLatest,
  select
} from 'redux-saga/effects'
import { cookieStorageMechanism, createClient, logger } from '~helpers'
import { getCustomerTokenName } from '../helpers'
import {
  getCustomerTrigger,
  getCustomerRequest,
  getCustomerSuccess,
  getCustomerFailure,
  getCustomerFulfill,
  triggerReadCustomer,
  readCustomer
} from '../slices'
import type { GetCustomerTriggerPayload, Customer } from '../types'
import { getErrorMessage } from '~helpers/error'
import { selectIsLoading } from '~domains/account/selectors/customer'

function* doGetCustomerSaga({
  payload
}: PayloadAction<GetCustomerTriggerPayload>) {
  try {
    const isLoading: boolean = yield select(selectIsLoading)

    if (!isLoading) {
      yield put(getCustomerRequest())
      const locale: string = yield getContext('locale')
      const client = createClient({ locale })
      const response: AxiosResponse<Customer> = yield call(
        client.get,
        `/api/account/get-customer?customerRef=${payload.customerReference}`
      )
      yield put(getCustomerSuccess(response.data))
    }
  } catch (error) {
    logger.error(getErrorMessage(error))
    yield put(
      getCustomerFailure({
        message: getErrorMessage(error)
      })
    )
  } finally {
    yield put(getCustomerFulfill())
  }
}

function* doReadCustomerSaga(): SagaIterator {
  const storageMechanism = cookieStorageMechanism()
  const CUSTOMER_TOKEN_NAME = getCustomerTokenName()
  const customer = yield call(storageMechanism.get, CUSTOMER_TOKEN_NAME)

  if (customer) {
    const parsedCustomer = yield call(JSON.parse, customer)
    if (parsedCustomer?.ref) {
      yield put(
        readCustomer({
          hasPreviousSuccessfulOrder:
            parsedCustomer['has-previous-successful-order'],
          reference: parsedCustomer.ref
        })
      )
    }
  }
}

export function* getCustomerSaga(): SagaIterator {
  yield all([
    takeLatest(getCustomerTrigger, doGetCustomerSaga),
    takeLatest(triggerReadCustomer, doReadCustomerSaga)
  ])
}
