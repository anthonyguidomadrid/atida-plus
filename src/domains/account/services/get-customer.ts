import getConfig from 'next/config'
import { createClient } from '~helpers'
import { normalizeCustomer } from '../normalizers'
import {
  CustomerAddress,
  CustomerWithAddresses,
  SprykerCustomer
} from '../types'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const getCustomer = async (
  locale: string,
  customerReference: string,
  accessToken: string
): Promise<CustomerWithAddresses> => {
  const { serverRuntimeConfig } = getConfig()
  const client = createClient({
    locale,
    options: { baseURL: serverRuntimeConfig.ecommerceHost[locale] }
  })
  const response = await client.get<SprykerCustomer>(
    `/customers/${customerReference}`,
    {
      headers: {
        Authorization: `Bearer ${accessToken}`
      }
    }
  )

  const customer = normalizeCustomer(response.data)

  // TODO: needs test to ensure when attributes not present that this error is thrown
  if (!customer) {
    throw new Error('account.get-customer.failed-to-retrieve-customer-details')
  }

  return {
    ...customer,
    addresses:
      response.data?.data?.relationships?.addresses?.data
        ?.map(({ id }) => {
          const addressData = response?.data?.included?.find(
            address => id === address.id
          )
          return { ...addressData?.attributes, id: addressData?.id }
        })
        ?.filter((address): address is CustomerAddress => !!address) ?? []
  }
}
