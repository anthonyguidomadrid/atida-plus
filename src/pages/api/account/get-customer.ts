import type { NextApiRequest, NextApiResponse } from 'next'
import { getCustomer } from '~domains/account/services/get-customer'
import { getJWTName } from '~domains/account'
import { handleApiError, isApiError, handleUnknownError } from '~helpers/error'
import { getDefaultLocale } from '~domains/translated-routes'

export default async (
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> => {
  try {
    const locale = req.headers['accept-language'] ?? getDefaultLocale()
    const data = req.query
    const token = req.cookies[getJWTName()]
    const result = await getCustomer(
      locale.toString(),
      String(data.customerRef),
      String(token)
    )
    res.statusCode = 200
    res.json(result)
  } catch (error) {
    isApiError(error)
      ? handleApiError(res, error, 'account.get-customer.error')
      : handleUnknownError(
          req,
          res,
          error,
          'account.get-customer.unexpected-error'
        )
  }
}
