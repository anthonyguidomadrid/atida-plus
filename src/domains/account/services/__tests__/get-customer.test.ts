/**
 * @jest-environment node
 */
import axios from 'axios'
import { createClient } from '~helpers'

import {
  sprykerCustomer,
  sprykerCustomerWithAddresses
} from '../../__mocks__/customer'
import { getCustomer } from '../get-customer'

describe(getCustomer, () => {
  describe('customer without addresses', () => {
    beforeEach(() => {
      ;(axios.get as jest.Mock).mockResolvedValueOnce({
        data: sprykerCustomer
      })
    })

    it('creates the client & passes the locale to spryker request', async () => {
      await getCustomer('en-gb', 'some-ref', 'some-token')
      expect(createClient).toHaveBeenCalledWith({
        locale: 'en-gb',
        options: {
          baseURL: 'spryker.co.uk'
        }
      })
      expect(axios.get).toHaveBeenCalledWith(`/customers/some-ref`, {
        headers: { Authorization: 'Bearer some-token' }
      })
    })

    it('returns the correct response', async () => {
      const getCustomerResponse = await getCustomer(
        'en-gb',
        'some-ref',
        'some-token'
      )
      expect(getCustomerResponse).toMatchSnapshot()
    })
  })

  describe('customer with addresses', () => {
    beforeEach(() => {
      ;(axios.get as jest.Mock).mockResolvedValueOnce({
        data: sprykerCustomerWithAddresses
      })
    })

    it('returns the correct response', async () => {
      const getCustomerResponse = await getCustomer(
        'en-gb',
        'some-ref',
        'some-token'
      )
      expect(getCustomerResponse).toMatchSnapshot()
    })
  })

  describe('incorrect customer response from Spryker with no attributes', () => {
    beforeEach(() => {
      ;(axios.get as jest.Mock).mockResolvedValueOnce({
        data: {
          data: {
            type: 'customers',
            id: 'PT--1821'
          }
        }
      })
    })
    it('Throws an error with the correct message', async () => {
      expect.assertions(1)
      await getCustomer('en-gb', 'some-ref', 'some-token').catch(e =>
        expect(e).toEqual(
          new Error('account.get-customer.failed-to-retrieve-customer-details')
        )
      )
    })
  })
})
