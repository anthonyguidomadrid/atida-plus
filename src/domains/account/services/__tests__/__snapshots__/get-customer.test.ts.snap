// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCustomer customer with addresses returns the correct response 1`] = `
Object {
  "accountType": "Personal",
  "addresses": Array [
    Object {
      "address1": "some-address",
      "address2": "some-address2",
      "address3": "some-address3",
      "city": "some-city",
      "company": "some-company",
      "country": "some-country",
      "firstName": "some-name",
      "id": "address-id-1",
      "isDefaultBilling": true,
      "isDefaultShipping": true,
      "isTaxExempt": false,
      "iso2Code": "some-iso-code",
      "lastName": "some-last-name",
      "phone": "some-phone",
      "salutation": "some-salutation",
      "zipCode": "some-zip-code",
    },
    Object {
      "address1": "some-other-address",
      "address2": "some-other-address2",
      "address3": "some-other-address3",
      "city": "some-other-city",
      "company": "some-other-company",
      "country": "some-other-country",
      "firstName": "some-other-name",
      "id": "address-id-2",
      "isDefaultBilling": false,
      "isDefaultShipping": false,
      "isTaxExempt": false,
      "iso2Code": "some-other-iso-code",
      "lastName": "some-other-last-name",
      "phone": "some-other-phone",
      "salutation": "some-other-salutation",
      "zipCode": "some-other-zip-code",
    },
  ],
  "company": null,
  "email": "some@email.com",
  "emailNotification": true,
  "firstName": "Test",
  "lastName": "McTest",
  "salutation": "Mr",
  "surcharge": false,
}
`;

exports[`getCustomer customer without addresses returns the correct response 1`] = `
Object {
  "accountType": "Personal",
  "addresses": Array [],
  "company": null,
  "email": "some@email.com",
  "emailNotification": true,
  "firstName": "Test",
  "lastName": "McTest",
  "salutation": "Mr",
  "surcharge": false,
}
`;
